kernel/src/hil/framebuffer.rs

```diff
@@ -0,0 +1,169 @@
+    /// note that width and height may change due to rotation
+    /// This function is synchronous as the driver should know this value without
+    /// requesting it from the screen.
+    fn get_supported_resolution(&self, index: usize) -> Option<(usize, usize)>;
+
+    /// Returns the number of the pixel formats supported.
+    /// This function is synchronous as the driver should know this value without
+    /// requesting it from the screen (most screens do not support such a request,
+    /// pixel formats are described in the data sheet).
+    ///
+    /// If the screen supports such a feature, the driver should request this information
+    /// from the screen upfront.
+    fn get_supported_pixel_formats(&self) -> usize;
+
+    /// Can be called with index 0 .. count-1 and will returns
+    /// the value of each pixel format mode.
+    /// This function is synchronous as the driver should know this value without
+    /// requesting it from the screen.
+    fn get_supported_pixel_format(&self, index: usize) -> Option<ScreenPixelFormat>;
+}
+
+pub trait Screen {
+    /// Returns a tuple (width, height) with the current resolution (in pixels)
+    /// This function is synchronous as the driver should know this value without
+    /// requesting it from the screen.
+    ///
+    /// note that width and height may change due to rotation
+    fn get_resolution(&self) -> (usize, usize);
+
+    /// Returns the current pixel format
+    /// This function is synchronous as the driver should know this value without
+    /// requesting it from the screen.
+    fn get_pixel_format(&self) -> ScreenPixelFormat;
+
+    /// Returns the current rotation.
+    /// This function is synchronous as the driver should know this value without
+    /// requesting it from the screen.
+    fn get_rotation(&self) -> ScreenRotation;
+
+    /// Sends a write command to write data in the selected video memory window.
+    /// The screen will then call ``ScreenClient::fill_next_buffer_for_write`` for
+    /// the actual bytes to write. This function will fill the buffer  and return
+    /// the number of bytes written. If it returns 0, the write stops and the screen
+    /// issues ``ScreenClient::command_complete``.
+    /// This avoids triple buffering (an app buffer, a frame buffer buffer and a screen buffer),
+    /// data is transfered from the app directly from the AppShare.
+    /// This also allows writing a repeated pattern with the app only having to fill a buffer
+    /// with one repeated sample. It also allow the screen to have
+    /// an internal arbitrary size buffer.
+    fn write(&self, x: usize, y: usize, width: usize, height: usize) -> ReturnCode;
+
+    fn set_client(&self, client: Option<&'static dyn ScreenClient>);
+
+    /// Powers up the display.
+    fn on(&self) -> ReturnCode;
+
+    /// Powers down the display. The screen should be able to accept data even when the display is off.
+    fn off(&self) -> ReturnCode;
```

Comment:
What about setting a brightness level instead? Something between 0 and 1 maybe, where 0 is off?

Reply:
Most of the embedded screens have the brightness set as a background LED. I could maybe add that function in the ScreenSetup (advanced) HIL?

Reply:
If the hardware only supports on/off then the brightness could maybe just be rounded up/down. I think for hardware that supports multiple brightness levels (if it exists) if would be useful to set a brightness.

Reply:
Done, thanks for the suggestions. I used a value in percent, as sending float values is not trivial.

Reply:
I made the following changes to the HIL so that the buffer is sent from the frame buffer driver towards the lower level screen driver:

```rust
/// Sets the video memory frame.
fn set_write_frame(&self, x: usize, y: usize, width: usize, height: usize) -> ReturnCode;

/// Sends a write command to write data in the selected video memory frame.
fn write(&self, buffer: &'static mut [u8], len: usize) -> ReturnCode;
```

I added another function to the ScreenClient to signal when a write command is complete.

```rust
/// The screen will call this function to notify that a command (except write) has finished.
fn command_complete(&self, r: ReturnCode);

/// The screen will call this function to notify that the write command has finished.
/// This is different from `command_complete` as it has to pass back the write buffer
fn write_complete(&self, buffer: &'static mut [u8], r: ReturnCode);
```
Any feedback is very useful.